Sources/Diffing.swift

```diff
@@ -158,3 +148,147 @@ extension CollectionSectionViewModel: DifferentiableSection {
+
+    /// Static diffing key used when diffing off-screen models
+    private static let placeholderDiffingKey = UUID().uuidString
+
+    /// When true, we allow diffing to access the real model's diffing key, eagerly loading it
+    private let _inVisibleBounds: Bool
+
+    /// Lambda to load the model
+    private let _modelGetter: () -> TableCellViewModel
+
+    /// Lazy reference to the model
+    private lazy var model = self._modelGetter()
+
+    init(inVisibleBounds: Bool, modelGetter: @escaping () -> TableCellViewModel) {
+        self._inVisibleBounds = inVisibleBounds
+        self._modelGetter = modelGetter
+    }
+
+    /// Only accessed during display, so eager-loading is allowed here
+    var accessibilityFormat: CellAccessibilityFormat {
+        self.model.accessibilityFormat
+    }
+
+    /// Only called during display, so eager-loading is allowed here
+    func applyViewModelToCell(_ cell: UITableViewCell) {
+        self.model.applyViewModelToCell(cell)
+    }
+
+    /// Only accessed during display, so eager-loading is allowed here
+    var registrationInfo: ViewRegistrationInfo {
+        self.model.registrationInfo
+    }
+
+    /// Only allows accessing the real model's diffing key for visible models
+    var diffingKey: DiffingKey {
+        if self._inVisibleBounds {
+            return self.model.diffingKey
+        } else {
+            return Self.placeholderDiffingKey
+        }
+    }
+}
+
+/// A `DifferentiableSection` that ensures we only allow eager-loading
+/// of cells that are known to be on-screen, which avoids forcing
+/// a datasource to potentially load data that hasn't been loaded yet
+/// to create new cell models (expensive)
+struct DiffableTableSectionViewModel: Collection, DifferentiableSection {
+
+    /// Reference to the original `TableSectionViewModel`
+    fileprivate let _sectionModel: TableSectionViewModel
+
+    /// The set of indices in this section, for which we should allow
+    /// diffing to access, since these cells are visible (i.e. won't
+    /// eagerly load data that the data source may not have loaded)
+    private let _visibleIndices: Set<Int>
+
+    /// Initializes a `DiffableTableSectionViewModel` with the
+    /// section model and visibile indices in this section
+    init(sectionModel: TableSectionViewModel, visibleIndices: Set<Int>) {
+        self._sectionModel = sectionModel
+        self._visibleIndices = visibleIndices
+    }
+
+    /// MARK: - Protocol Implementations
+
+    /// :nodoc:
+    init<C: Swift.Collection>(source: DiffableTableSectionViewModel, elements: C) where C.Element == AnyDiffableViewModel {
+        self._sectionModel = TableSectionViewModel(
+            diffingKey: source._sectionModel.diffingKey,
+            cellViewModelDataSource: TableCellViewModelDataSource(
+                //swiftlint:disable:next force_cast
+                elements.map { $0.model as! TableCellViewModel }
```

Comment:
Might be nice to add some documentation here about why this force cast is acceptable - isn't `model` not necessarily guaranteed to be a TableCellViewModel?

Reply:
it is in this case. i'll document